src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -390,28 +495,25 @@ internal static void Unescape(ReadOnlySpan<byte> source, Span<byte> destination,
 
                             Debug.Assert(JsonHelpers.IsInRangeInclusive((uint)scalar, JsonConstants.HighSurrogateStartValue, JsonConstants.HighSurrogateEndValue));
 
-                            idx += 3;   // Skip the last hex digit and the next \u
-
                             // We must have a low surrogate following a high surrogate.
-                            if (source.Length < idx + 4 || source[idx - 2] != '\\' || source[idx - 1] != 'u')
+                            if (source.Length < idx + 7 || source[idx + 1] != '\\' || source[idx + 2] != 'u')
                             {
-                                ThrowHelper.ThrowInvalidOperationException_ReadInvalidUTF16();
+                                ThrowHelper.ThrowInvalidOperationException_ReadIncompleteUTF16();
                             }
 
                             // The source is known to be valid JSON, and hence if we see a \u, it is guaranteed to have 4 hex digits following it
                             // Otherwise, the Utf8JsonReader would have already thrown an exception.
-                            result = Utf8Parser.TryParse(source.Slice(idx, 4), out int lowSurrogate, out bytesConsumed, 'x');
+                            result = Utf8Parser.TryParse(source.Slice(idx + 3, 4), out int lowSurrogate, out bytesConsumed, 'x');
                             Debug.Assert(result);
                             Debug.Assert(bytesConsumed == 4);
+                            idx += 6;
```

Comment:
can you add a comment why 6 and not 7? Previously there was `The loop iteration will increment idx past the last hex digit`

Reply:
That's no longer the case since I changed the iteration from a for loop to a `while(true)` loop.

Reply:
so why 6? You seem to increment by 3, then consume 4 and you somehow end up with 6?

Reply:
The format is `\u` followed by 4 hex digits, for a total of 6 characters.

Reply:
but `\uXYVZ` starts at `source[idx + 1]` at least that's what you check on line 499. I think this should at least have some comment

Reply:
It's the same invariant as with all other escape sequences (the index is incremented at the end of the loop). Any comment should be applied at the base of loop block, but it's pretty self explanatory once you look at how the loop works. FWIW this is not an invariant my changes introduced.